Parameterize `hvm_disk_image` variable

nodeg opened the pull request:
## What does this PR change?

This will add the possibility to parameterize the `hvm_disk_image` variable and use more than 1 image. 

### TODO before merging

- [x] Check CI pipelines `main.tf` if they use the changes variable: https://github.com/SUSE/susemanager-ci/search?q=hvm_disk_image
- [x] Adjust files in `susemanager-ci`: https://github.com/SUSE/susemanager-ci/pull/709
- [x] Expand the README
- [x] simplify the variables and create a `map(map(string))` instead of multiple `map(string)` if possible
- [x] only leave the default vor openSUSE Leap which is publicly available and remove the SLES part
- [x] Rebase and squash commits


See https://github.com/SUSE/spacewalk/issues/17238

rjmateus commented on modules/virthost/main.tf:
```diff
@@ -19,8 +19,10 @@ module "virthost" {
   additional_grains = merge({
-    hvm_disk_image      = var.hvm_disk_image
-    hvm_disk_image_hash = var.hvm_disk_image_hash
+    hvm_disk_image      = var.hvm_disk_image["leap"]
```
I agree with Cedric here, we could just pass the map to gains, and adapt the sls accordingly.
If you prefer to keep it is fine, but we may need to refactor it in the future if more templates need to be added.

nodeg replied:
I would keep it for now and adjust it when adding the Debian-like and Red Hat-like images later on.

nodeg replied:
What do you think @cbosdo?

cbosdo replied:
As you like... I would prefer doing it right from the start, but you are the one hacking it

cbosdo replied:
Be aware also that the `leap` image here is any HVM image and could be a SLES as well for the testsuite. You probably want to clarify this in the cucumber KVM scenario

cbosdo replied:
you still need to get the grain as a list here

cbosdo commented on modules/virthost/variables.tf:
```diff
@@ -80,13 +80,21 @@ variable "ipv6" {
-  default     = "https://download.opensuse.org/distribution/leap/15.3/appliances/openSUSE-Leap-15.3-JeOS.x86_64-OpenStack-Cloud.qcow2.sha256"
+  description = "Hash of the HVM disk images, either a URL or the hash itself. See salt's file.managed source_hash documentations"
+  type = map(string)
+  default = {
```
You probably want to check the CI pipelines `main.tf` files and change them if they used that variable.

nodeg replied:
Normally I would also prefer doing this in the first place, but I am not that familiar with Salt states. Do you have an example on how to integrate that map into the `.sls` file?

rjmateus replied:
here is an example: https://github.com/uyuni-project/sumaform/blob/master/salt/default/network.sls#L13-L18

nodeg replied:
Thank you Ricardo, I will have a look at them.

nodeg commented:
@cbosdo or @rjmateus is is possible to get the FQDN/honstname from the server to insert it into my [cloud init config](https://github.com/uyuni-project/sumaform/pull/1197/files#diff-a64442233c25b8789d1acbf5f8932897e60a55bd412d84a3beb4c967ed714757R25)?

```sls
# adjust configuration files
write_files:
- content: |
   enable_legacy_startup_events: False
   enable_fqdns_grains: False
   master: # TODO FQDN server
  path: /etc/salt/minion
```
I need this for onboarding my nested VM [via the Salt OS](https://documentation.suse.com/suma/4.3/en/suse-manager/client-configuration/registration-cli.html#_salt_minion_configuration) and not the bundle.

cbosdo commented:
> @cbosdo or @rjmateus is is possible to get the FQDN/honstname from the server to insert it into my [cloud init config](https://github.com/uyuni-project/sumaform/pull/1197/files#diff-a64442233c25b8789d1acbf5f8932897e60a55bd412d84a3beb4c967ed714757R25)?
> 
> ```yaml-salt
> # adjust configuration files
> write_files:
> - content: |
>    enable_legacy_startup_events: False
>    enable_fqdns_grains: False
>    master: # TODO FQDN server <-----------
>   path: /etc/salt/minion
> ```
> 
> I need this for onboarding my nested VM [via the Salt OS](https://documentation.suse.com/suma/4.3/en/suse-manager/client-configuration/registration-cli.html#_salt_minion_configuration) and not the bundle.

You can use `{{grains['server']}}` to get the server FQDN from the Salt states.

nodeg commented on salt/virthost/cloudinit/user-data:
```diff
@@ -0,0 +1,64 @@
+   enable_fqdns_grains: False
+  path: /etc/salt/minion
+- content: |
+   salt-sles-migration-test.tf.local
+  path: /etc/hostname
+- content: |
+   [server]
+   domain-name=tf.local
+   use-ipv4=yes
+   use-ipv6=no
+   ratelimit-interval-usec=1000000
+   ratelimit-burst=1000
+   [wide-area]
+   enable-wide-area=yes
+   [publish]
+   publish-hinfo=no
+   publish-workstation=no
+  path: /etc/avahi/avahi-daemon.conf
+- content: |
+   .local
+   .tf.local
+  path: /etc/mdns.allow
+- content: |
+   passwd:         compat
+   group:          compat
+   shadow:         compat
+   hosts:          files mdns [NOTFOUND=return] dns
+   networks:       files dns
+   aliases:        files usrfiles
+   ethers:         files usrfiles
+   gshadow:        files usrfiles
+   netgroup:       files nis
+   protocols:      files usrfiles
+   publickey:      files
+   rpc:            files usrfiles
+   services:       files usrfiles
+   automount:      files nis
+   bootparams:     files
+   netmasks:       files
+  path: /etc/nsswitch.conf
```
Apparently only some of these changes were adapted in the VM.
Adjusted correctly:

- `/etc/salt/minion`
- `/etc/mdns.allow`

Not adjusted:

- `/etc/avahi/avahi-daemon.conf`
- `/etc/nsswitch.conf`

@cbosdo do you have any idea what the issue could be?

nodeg commented:
Ok. I tried several things, did several sumaform deployments and was not able to get it the way I wanted to be so I will leave it this way and maybe I or someone else can have a look in the future and improve/edit the way I did it now. I do not want to spent more time on that at the moment.
@cbosdo I had a look at the `main.tf` files in the [CI that use this variable](https://github.com/SUSE/susemanager-ci/search?q=hvm_disk_image). Since I did not change the name of the `hvm_disk_image`, the CI should be fine. I did rename the `qcow2` image file though but also renamed it in the KVM test in the testsuite: https://github.com/uyuni-project/uyuni/pull/5240/commits/d1340ab3e1731532063387802da71413867a9037
I will squash all commits before merging.
I would be glad to have another review from you @cbosdo @rjmateus.

srbarrios reviewed (commented):
It looks good to me, but I would let Cedric and Ricardo to decide when this can be merged.

nodeg commented on salt/virthost/cloudinit/user-data:
```diff
@@ -0,0 +1,72 @@
+# adjust configuration files
+write_files:
+- content: |
+   master: {{grains['server']}}
```
I did another local sumaform deployment to test this. Apparently this will not get replaced correctly. We run this code on the virthost and not on the server. Could that be the issue? Any suggestions?

cbosdo replied:
when was this state executed? during the sumaform process or after registering the virtual host to the server?

cbosdo replied:
Why are you having all that logic into the cloud init? Smells bad to me 🙊 Couldn't this be done in the testsuite itself? Like Ricardo mentioned earlier I would keep cloud init generic enough to be reusable by other tests

nodeg replied:
This gets executed during the sumaform process.

cbosdo commented on salt/virthost/cloudinit/README.md:
```diff
@@ -0,0 +1,17 @@
+
+## Links
+
+For mor information see the official [cloud init docs](https://cloudinit.readthedocs.io/en/latest/topics/examples.html).
```
```suggestion
For more information see the official [cloud init docs](https://cloudinit.readthedocs.io/en/latest/topics/examples.html).
```

cbosdo commented on salt/virthost/cloudinit/user-data:
```diff
@@ -0,0 +1,72 @@
+  list: |
+     root:linux
+ssh_authorized_keys:
+- ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDOSLK0kJAulwO9Qwm/DVKLsGzv8Fh+ZcGe5x5d+VzO4/+xKnxjOPvGVngm9Hki8N436bCCvIRqwzstDUyJnTHhEe3qzsl02QGqIIfVjzK3BEp4sA8XcmByBy3pxqbTjAjrW8a+n43DBRMx4J86LgywX3fFt8ceEym3rFer/IzQD9QvKRocycNCwhpmGKEDh4zdl9PdyU5RcUpiA01AD/8rdJMNYnUxIsaNElcgP2LSzROjNdyu11cq9h3/cb3b7oRRef718GOkIkzqL/szo4Mm/j4KfyG0SBWyPxlJktFADD3lcQIRrGakFhiZOFbg2HZci26q+l/hUfLMbzcBrWlc+gzY4z4LMF8YBy1x/ ...
```
Is it OK to hard-code this SSH key here? I believe this should be replaced by a grain value

nodeg replied:
The [example](https://cloudinit.readthedocs.io/en/latest/topics/examples.html#including-users-and-groups) in the cloud init docs does the same. Nevertheless I can have a look how to replace that.
There is e.g. `salt.states.file.append` but I cannot specify a location inside the file where I want to append things which is important in my case.

cbosdo commented on salt/virthost/cloudinit/user-data:
```diff
@@ -0,0 +1,72 @@
+  path: /etc/hostname
+- content: |
+   [server]
+   domain-name=tf.local
```
This is configurable in sumaform. There surely is a grain for that

cbosdo commented on salt/virthost/cloudinit/user-data:
```diff
@@ -0,0 +1,72 @@
+   enable_fqdns_grains: False
+  path: /etc/salt/minion
+- content: |
+   sles-salt-migration.tf.local
```
Strange to see a hard-coded host name here... I think this should be in a grain as this could change depending on the sumaform config. For instance I don't use `tf.local` as domain and the CI probably doesn't either.

cbosdo commented on salt/virthost/init.sls:
```diff
@@ -77,6 +77,55 @@ disk-image-template.qcow2:
+
+create-vm-cloudinit-disk:
+  cmd.run:
+    - name: mkisofs -o /var/testsuite-data/cloudinit-disk.iso -volid cidata -joliet -rock /root/cloudinit
```
Using `/root` for this is probably not a super good idea... but should work

nodeg replied:
Fixed.

cbosdo reviewed (commented):
Raising more questions than you asked... sorry about that

nodeg commented:
Most of the logic has to be in the cloud init config so when the test suite sets up the nested VM we should have

- a hostname/FQDN
- correct configuration for Avahi, Salt, DNS, etc.

otherwise the VM cannot be seen/onboarded.

I mean I can move that from the cloud init configs to a Salt state to make it configurable via the sumaform configuration but it has to be set up before we ran the test suite.

nodeg commented:
Ok, I moved all the logic into the Salt state file for the `virthost`. The replacement of the grains I am querying does also work.

nodeg commented:
I adjusted the variables and tested a local sumaform deployment.

I furthermore took a look at `susemanger-ci`. The minimal adjustments for e.g. [4.2](https://github.com/SUSE/susemanager-ci/blob/a84c9b240471e1913f28bfefdcd648b513a59ec3/terracumber_config/tf_files/SUSEManager-4.2-PRV.tf#L192-L193) would be:


**terracumber_config/tf_files/SUSEManager-4.2-PRV.tf**:
```diff
    kvm-host = {
      image = "sles15sp3o"
      additional_grains = {
-        hvm_disk_image = "http://minima-mirror.mgr.prv.suse.net/distribution/leap/15.3/appliances/openSUSE-Leap-15.3-JeOS.x86_64-OpenStack-Cloud.qcow2"
-        hvm_disk_image_hash = "http://minima-mirror.mgr.prv.suse.net/distribution/leap/15.3/appliances/openSUSE-Leap-15.3-JeOS.x86_64-OpenStack-Cloud.qcow2.sha256"
+        hvm_disk_image = {
+         leap = "http://minima-mirror.mgr.prv.suse.net/distribution/leap/15.3/appliances/openSUSE-Leap-15.3-JeOS.x86_64-OpenStack-Cloud.qcow2"
+         sles = "https://download.suse.de/install/SLE-15-SP4-Minimal-GM/SLES15-SP4-Minimal-VM.x86_64-OpenStack-Cloud-GM.qcow2"
+        }
+        hvm_disk_image_hash = {
+         leap = "http://minima-mirror.mgr.prv.suse.net/distribution/leap/15.3/appliances/openSUSE-Leap-15.3-JeOS.x86_64-OpenStack-Cloud.qcow2.sha256"
+         sles = "1215677036ea32bf2a8d99770b28cc0e76956def0c14d7a8d30f745ec1f0438c"
+        }
      }
```
Would this be enough and is this the correct way to do it @cbosdo?

The other 2 variables I introduced to the virthost is `hvm_disk_image_name` and `hvm_disk_image_hostname` which do have [default values](https://github.com/uyuni-project/sumaform/pull/1197/files#diff-1a1c4bf64e79a8f4155214c532765cfe62ffcd67e8e3b5a4da7df0bf157944f0) specified, so FMPOV they do not have to be explicitly specified in the `susemanager-ci` terraform files.

```hcl
variable "hvm_disk_image_name" {
  description = "Name of the HVM disk images used as key in other variables"
  type = map(string)
  default = {
    leap = "leap"
    sles = "sles"
  }
}

variable "hvm_disk_image_hostname" {
  description = "Hostname of the HVM disk images"
  type = map(string)
  default = {
    leap = "leap153"
    sles = "sles-salt-migration"
  }
}
```

cbosdo commented:
> I adjusted the variables and tested a local sumaform deployment.
> 
> I furthermore took a look at `susemanger-ci`. The minimal adjustments for e.g. [4.2](https://github.com/SUSE/susemanager-ci/blob/a84c9b240471e1913f28bfefdcd648b513a59ec3/terracumber_config/tf_files/SUSEManager-4.2-PRV.tf#L192-L193) would be:
> 
> **terracumber_config/tf_files/SUSEManager-4.2-PRV.tf**:
> 
> ```diff
>     kvm-host = {
>       image = "sles15sp3o"
>       additional_grains = {
> -        hvm_disk_image = "http://minima-mirror.mgr.prv.suse.net/distribution/leap/15.3/appliances/openSUSE-Leap-15.3-JeOS.x86_64-OpenStack-Cloud.qcow2"
> -        hvm_disk_image_hash = "http://minima-mirror.mgr.prv.suse.net/distribution/leap/15.3/appliances/openSUSE-Leap-15.3-JeOS.x86_64-OpenStack-Cloud.qcow2.sha256"
> +        hvm_disk_image = {
> +         leap = "http://minima-mirror.mgr.prv.suse.net/distribution/leap/15.3/appliances/openSUSE-Leap-15.3-JeOS.x86_64-OpenStack-Cloud.qcow2"
> +         sles = "https://download.suse.de/install/SLE-15-SP4-Minimal-GM/SLES15-SP4-Minimal-VM.x86_64-OpenStack-Cloud-GM.qcow2"
> +        }
> +        hvm_disk_image_hash = {
> +         leap = "http://minima-mirror.mgr.prv.suse.net/distribution/leap/15.3/appliances/openSUSE-Leap-15.3-JeOS.x86_64-OpenStack-Cloud.qcow2.sha256"
> +         sles = "1215677036ea32bf2a8d99770b28cc0e76956def0c14d7a8d30f745ec1f0438c"
> +        }
>       }
> ```
> 
> Would this be enough and is this the correct way to do it @cbosdo?

Seems good to me. I'm not an HCL expert though 😉 

> The other 2 variables I introduced to the virthost is `hvm_disk_image_name` and `hvm_disk_image_hostname` which do have [default values](https://github.com/uyuni-project/sumaform/pull/1197/files#diff-1a1c4bf64e79a8f4155214c532765cfe62ffcd67e8e3b5a4da7df0bf157944f0) specified, so FMPOV they do not have to be explicitly specified in the `susemanager-ci` [terraform files](https://github.com/SUSE/susemanager-ci/pull/709).
> 
> ```terraform
> variable "hvm_disk_image_name" {
>   description = "Name of the HVM disk images used as key in other variables"
>   type = map(string)
>   default = {
>     leap = "leap"
>     sles = "sles"
>   }
> }

That one looks funny... Can't it be simplified?

> variable "hvm_disk_image_hostname" {
>   description = "Hostname of the HVM disk images"
>   type = map(string)
>   default = {
>     leap = "leap153"
>     sles = "sles-salt-migration"
>   }
> }
> ```

OK

nodeg commented:
Thank you for the comments @cbosdo. Do you have a recommendation for a simplification?

nodeg commented:
I found one more issue while testing this manually.

```bash
sles-salt-migration:~ # cat /etc/salt/minion
master: sumaform-pxy.tf.local
server_id_use_crc: adler32
enable_legacy_startup_events: False
enable_fqdns_grains: False
```
The `master` is set to the proxy and not the server although I use the following:

```salt
        - content: |
           master: {{ grains.get('server') }}
           server_id_use_crc: adler32
           enable_legacy_startup_events: False
           enable_fqdns_grains: False
          path: /etc/salt/minion
```
When querying this grain on the virthost Minion the server grain is indeed set to the proxy. 

```bash
sumaform-min-kvm:/var/testsuite-data # salt-call --local grains.item server
local:
    ----------
    server:
        sumaform-pxy.tf.local
```

Is this intentional?

cbosdo commented:
> Thank you for the comments @cbosdo. Do you have a recommendation for a simplification?

Can't you just suppose and document the variables are consistent across variables? or even document what are the expected values and where it should match?

cbosdo commented:
> Is this intentional? It does work to onboard the Minion, though. Just asking.

When the proxy is configured every thing is set to go through it in the testsuite. So I think this is intentional

Bischoff commented on README_ADVANCED.md:
```diff
@@ -231,12 +231,14 @@ The generated virtual host will be setup with:
+- and a VM template disk image located in `/var/testsuite-data/`.
 
-The template disk image is the `opensuse153` image used by sumaform and is downloaded when applying the highstate on the virtual host.
+The openSUSE Leap template (`leap`) disk image is `opensuse153o` used by sumaform, the SLES template (`sles`) is a
```
Make it clear when we use the leap one and when we use the SLES one ?

Bischoff replied:
is the above wrong?
```
SUSE/susemanager-ci/terracumber_config/tf_files> grep hvm SUSEManager-*-PRV.tf 
SUSEManager-4.2-PRV.tf:        hvm_disk_image =
 "http://minima-mirror.mgr.prv.suse.net/distribution/leap/15.3/appliances/openSUSE-Leap-15.3-JeOS.x86_64-OpenStack-Cloud.qcow2"
SUSEManager-4.2-PRV.tf:        hvm_disk_image_hash =
 "http://minima-mirror.mgr.prv.suse.net/distribution/leap/15.3/appliances/openSUSE-Leap-15.3-JeOS.x86_64-OpenStack-Cloud.qcow2.sha256"
SUSEManager-4.3-PRV.tf:        hvm_disk_image =
 "http://minima-mirror.mgr.prv.suse.net/distribution/leap/15.4/appliances/openSUSE-Leap-15.4-JeOS.x86_64-OpenStack-Cloud.qcow2"
SUSEManager-4.3-PRV.tf:        hvm_disk_image_hash =
 "http://minima-mirror.mgr.prv.suse.net/distribution/leap/15.4/appliances/openSUSE-Leap-15.4-JeOS.x86_64-OpenStack-Cloud.qcow2.sha256"
```
it looks like we have both Leap 15.3 and 15.4

(btw, is there a reason for keeping 15.3?)

nodeg replied:
I do not want to add more to this PR. Updating Leap to 15.4 can be done in a separate PR.

I am not sure if I should add more information about our test suite to the general sumaform docs here. I can explain it to you here if that is enough:

- Leap is used for the general KVM test in our test suite
- SLES is used for the Salt bundle migration test

Bischoff replied:
1) ok   2) ok, fair enough

Bischoff commented on README_TESTING.md:
```diff
@@ -292,16 +292,24 @@ User may need to change the KVM/Xen image download. To do it, one can use the `a
+            hvm_disk_image_hash = {
+              leap = "..."
+            }
             xen_disk_image = "..."
```
does it make sense to have xen data on a kvm host?

cbosdo replied:
the virthost module was used for both KVM and Xen. I believe we can now remove all Xen support from sumaform, but that could be done in a separate PR

Bischoff commented on README_TESTING.md:
```diff
@@ -292,16 +292,24 @@ User may need to change the KVM/Xen image download. To do it, one can use the `a
         additional_grains = {
-            hvm_disk_image = ".."
-            hvm_disk_image_hash = "..."
+            hvm_disk_image = {
```
does it make sense to have kvm data on a xen host?

Bischoff reviewed (commented):
see nitpicks

Bischoff reviewed (approved):
we might need to do other things in other PRs:
* remove xen
* move from leap 15.3 to 15.4

approving

nodeg commented:
For XEN I already created a card for hexagon. Updating Leap ist one thing I will do after the Salt bundle test PRs got merged.

nodeg commented:
Ok, so `map(map(string))` does work:

```salt
sumaform-min-kvm:~ # salt-call --local grains.item hvm_disk_image
local:
    ----------
    hvm_disk_image:
        ----------
        leap:
            ----------
            hash:
                https://download.opensuse.org/distribution/leap/15.3/appliances/openSUSE-Leap-15.3-JeOS.x86_64-OpenStack-Cloud.qcow2.sha256
            hostname:
                leap153
            image:
                https://download.opensuse.org/distribution/leap/15.3/appliances/openSUSE-Leap-15.3-JeOS.x86_64-OpenStack-Cloud.qcow2
            name:
                leap
        sles:
            ----------
            hash:
                http://minima-mirror.mgr.suse.de/install/SLE-15-SP4-Minimal-GM/SLES15-SP4-Minimal-VM.x86_64-OpenStack-Cloud-GM.qcow2.sha256
            hostname:
                sles-salt-migration
            image:
                https://download.suse.de/install/SLE-15-SP4-Minimal-GM/SLES15-SP4-Minimal-VM.x86_64-OpenStack-Cloud-GM.qcow2
            name:
                sles
```

nodeg commented:
I am currently testing this locally and will then request another review from you @cbosdo :wink:

nodeg commented:
Ok, sumaform is deploying successfully.

nodeg commented:
Apparently it did not work that well. I did a deployment where this branched was checked out. My `main.tf` for the virthost looks like the following:

```hcl
    kvm-host = {
      hvm_disk_image = {
        leap = {
          name = "leap"
          hostname = "leap-salt-migration"
          image = "http://minima-mirror.mgr.prv.suse.net/distribution/leap/15.4/appliances/openSUSE-Leap-15.4-JeOS.x86_64-OpenStack-Cloud.qcow2"
          hash = "http://minima-mirror.mgr.prv.suse.net/distribution/leap/15.4/appliances/openSUSE-Leap-15.4-JeOS.x86_64-OpenStack-Cloud.qcow2.sha256"
        }
        sles = {
          name = "sles"
          hostname = "sles-salt-migration"
          image = "http://minima-mirror.mgr.prv.suse.net/install/SLE-15-SP4-Minimal-GM/SLES15-SP4-Minimal-VM.x86_64-OpenStack-Cloud-GM.qcow2"
          hash = "http://minima-mirror.mgr.prv.suse.net/install/SLE-15-SP4-Minimal-GM/SLES15-SP4-Minimal-VM.x86_64-OpenStack-Cloud-GM.qcow2.sha256"
        }
      }
```
but when verifying this after the sumaform deployment there was only the default value from sumaform present on the virthost:

```bash
sumaform-min-kvm:~/salt/virthost # salt-call --local grains.item hvm_disk_image
local:
    ----------
    hvm_disk_image:
        ----------
        leap:
            ----------
            hash:
                https://download.opensuse.org/distribution/leap/15.3/appliances/openSUSE-Leap-15.3-JeOS.x86_64-OpenStack-Cloud.qcow2.sha256
            hostname:
                leap153
            image:
                https://download.opensuse.org/distribution/leap/15.3/appliances/openSUSE-Leap-15.3-JeOS.x86_64-OpenStack-Cloud.qcow2
            name:
                leap
sumaform-min-kvm:~/salt/virthost # 
```

cbosdo commented on salt/virthost/init.sls:
```diff
@@ -88,6 +74,148 @@ ifcfg-br0:
+### adjustments for cloud init and the test suite ---
+# https://cloudinit.readthedocs.io/en/latest/topics/examples.html
+{% for os_type in grains.get('hvm_disk_image') %}
+rezise-{{ salt['grains.get']('hvm_disk_image:' ~ os_type ~ ':name') }}-disk-image-template:
```
Why don't you use `os_type` directly instead of `salt['grains.get']('hvm_disk_image:' ~ os_type ~ ':name')` since those are basically the same in your values?

I would drop the `name` property and just use the map key instead: one less duplicate data

cbosdo commented on modules/virthost/variables.tf:
```diff
@@ -80,13 +80,16 @@ variable "ipv6" {
+  type = map(map(string))
+  default = {
+    leap = {
+      name = "leap"
```
You probably can get rid of the `name` value and use the item's key instead.

nodeg replied:
Fixed.

cbosdo commented on README_TESTING.md:
```diff
@@ -290,11 +290,17 @@ You may need to change the KVM image download. To do it, use the `additional_gra
+    additional_grains = {
+      hvm_disk_image = {
+        leap = {
+          name = "..."
```
`name` should be removed here too

nodeg replied:
Fixed.

cbosdo reviewed (commented):
Some more data deduplication is possible

vandabarata reviewed (approved):
lgtm, I'd just do the changes suggested by Cedric to avoid duplication

nodeg commented:
Rebased and adjusted the commits.

cbosdo commented on README_ADVANCED.md:
```diff
@@ -230,12 +230,14 @@ The generated virtual host will be setup with:
 In order to use another or a cached image, use the `hvm_disk_image` variable.
-If the `hvm_disk_image` is set to the empty string, no image will be copied in `/var/testsuite-data/`.
-For example, to use a local image copy it in `salt/virthost/` folder and set `hvm_disk_image = "salt://virthost/imagename.qcow2"`
+If the values inside the `hvm_disk_image` map are set to an empty string, no image will be copied to `/var/testsuite-data/`.
```
Empty map... is the testsuite automatically disabling the virtualization features if the image isn't there? I expect there could be issues with this..

nodeg replied:
No, the feature is not disabled but we raise a warning and abort the current scenario. The sentence above is to inform the user that if the values are empty, we do not have any images in the mentioned folder above.

cbosdo commented on salt/virthost/init.sls:
```diff
@@ -88,6 +74,148 @@ ifcfg-br0:
+
+### adjustments for cloud init and the test suite ---
+# https://cloudinit.readthedocs.io/en/latest/topics/examples.html
+{% for os_type in grains.get('hvm_disk_image') %}
```
How comes this is not merged with the previous loop since those are looping on the same things?

nodeg replied:
Nice catch. Before addressing your other suggestion regarding the name variable above it was not possible to use one loop. After changing the code, this for loop can be "merged" with the other. Thank you.

nodeg replied:
I amended the changes.

cbosdo reviewed (approved):
Left a couple of comments that could be addressed later
